book/src/platformer/03_importing_tiled_levels.md

```diff
@@ -0,0 +1,325 @@
+We'll not do this.
+
+# Some tiled boilerplate
+
+Working with the tiled library isn't ideal.
+For instance, we've used two layers that we've given nice names to, it would be nice to use these names to access the layers themselves.
+The library we will be using doesn't support this.
+So we will use the normal trick of making a trait that we implement on the foreign type.
+
+```rust
+trait GetLayer {
+    fn get_layer_by_name(&self, name: &str) -> Layer;
+    fn get_tile_layer(&self, name: &str) -> FiniteTileLayer;
+    fn get_object_layer(&self, name: &str) -> ObjectLayer;
+}
+
+impl GetLayer for Map {
+    fn get_layer_by_name(&self, name: &str) -> Layer {
+        self.layers().find(|x| x.name == name).unwrap()
+    }
+    fn get_tile_layer(&self, name: &str) -> FiniteTileLayer {
+        match self.get_layer_by_name(name).as_tile_layer().unwrap() {
+            TileLayer::Finite(finite_tile_layer) => finite_tile_layer,
+            TileLayer::Infinite(_) => panic!("Infinite tile layer not supported"),
+        }
+    }
+
+    fn get_object_layer(&self, name: &str) -> ObjectLayer {
+        self.get_layer_by_name(name).as_object_layer().unwrap()
+    }
+}
+```
+
+A `build.rs` file only runs when it changes or a dependency changes.
+What counts as a dependency?
+You have to tell `Cargo` each file you depend on by using [`rerun-if-changed`](https://doc.rust-lang.org/cargo/reference/build-scripts.html#rerun-if-changed).
+We can add this capability in the tiled library by using their `Reader` trait.
+
+```rust
+struct BuildResourceReader;
+
+impl ResourceReader for BuildResourceReader {
+    type Resource = <FilesystemResourceReader as ResourceReader>::Resource;
+
+    type Error = <FilesystemResourceReader as ResourceReader>::Error;
+
+    fn read_from(
+        &mut self,
+        path: &std::path::Path,
+    ) -> std::result::Result<Self::Resource, Self::Error> {
+        println!("cargo::rerun-if-changed={}", path.to_string_lossy());
+        FilesystemResourceReader.read_from(path)
+    }
+}
+```
+
+This adds a reader that passes through to the existing `FilesystemResourceReader` but intercepts each one to tell `cargo` to depend on the file being read.
+The whole reason for doing this is that loading a tiled map could mean we also need to load the various files it references, like the tilesets.
+If we were to change the tileset, maybe adding tiles or changing the tags, we would like that to be reflected in the next build of our game.
+Make sure to properly tell `cargo` about your dependencies as it will annoy you otherwise!
+
+
+# Loading the level into an internal representation
+
+It is best practice and easier to maintain to import the level into an internal representation and then convert that into your game representation.
+The internal representation can be inefficient as it's going to use your powerful build machine rather than the underpowered GBA.
```

Comment:
What do you mean by this?

Reply:
There are 4 possible tiles, how I have my intermediate representation inline the collision and win as separate bools is wildly inefficient.